diff -r 21f5af140f11 xml-rpc.el
--- a/xml-rpc.el	Thu Jan 15 20:03:23 2009 +0900
+++ b/xml-rpc.el	Thu Jan 15 20:06:43 2009 +0900
@@ -356,7 +356,7 @@
 	      (and (eq 1 (length charset-list))
 		   (eq 'ascii (car charset-list)))
 	      (not xml-rpc-base64-encode-unicode))
-	  `((value nil (string nil ,(url-insert-entities-in-string value))))
+	  `((value nil (string nil ,(xml-rpc-escape-if-need value))))
 	`((value nil (base64 nil ,(base64-encode-string
 				   (encode-coding-string value 'utf-8))))))))
    ((xml-rpc-value-doublep value)
@@ -667,7 +667,21 @@
 			  (stringp (car tree))))
 	    (insert ?\n indent-string))
 	  (insert ?< ?/ (symbol-name (xml-node-name xml)) ?>))))))
-    
+
+(defalias 'xml-rpc-escape-if-need
+  (if (string= (with-temp-buffer
+		 (xml-print '((a nil "A&B")))
+		 (buffer-string))
+	       "<a>A&B</a>")
+      ;; with old xml.el, need escape for output from xml-print<f>.
+      'url-insert-entities-in-string
+    ;; with new xml.el, already escaped by xml-print<f>
+    'identity)
+  "Insert entities in string made by old xml-print<f>.
+In old xml-print<f> does not escape entities in given string,
+but recent version does it. To be escaped or not to be is decided 
+on runtime by testing xml-print<f> behaviour.")
+
 (provide 'xml-rpc)
 
 ;;; xml-rpc.el ends here
